### PR TITLE
Simplify implementation of align_offset slightly

### PR DESCRIPTION
**Pull request by nagisa**

### CODE REVIEW

**Comment by SimonSapin** [comment]
I honestly have a hard time following the math in this code. Unfortunately it looks like the reviewers of the previous version in https://github.com/rust-lang/rust/pull/50319 did too :/

**Comment by nagisa** [comment]
I agree that the maths here are fairly complicated. It is fairly easy, however, to see that this PR keeps the code equivalence.

First, consider that the branch removed is `if gcd == 1 { ... }`. The branch after that, `if (p as usize & (gcd - 1) == 0) { … }`, if `gcd == 1`, simplifies as such: `if p & 0 == 0 { ... }` and thus will be taken whenever the original branch would have been.

To prove the equivalence between the branch bodies consider that `gcd = 1 << gcdpow` and for it to be `1`, `gcdpow` must be `0`.

In that case:

```
return intrinsics::unchecked_rem(a.wrapping_sub(pmoda).wrapping_mul(mod_inv(smoda, a)), a);
```

and

```
let j = a.wrapping_sub(pmoda) >> gcdpow; // same as `a.wrapping_sub(pmoda)`
let k = smoda >> gcdpow; // same as `smoda`
return intrinsics::unchecked_rem(j.wrapping_mul(mod_inv(k, a)), a >> gcdpow); 
```

are exactly the same.

---

The rest of the changes are documentation tweaks which I’ve figured as part of writing a blog post about some of the operations involved here (which I’m writing specifically with an intention to make the code here clearer). I’m willing to keep this PR open until I publish the post.

**Comment by nagisa** [comment]
I [blogged about the implementation](https://kazlauskas.me/entries/what-makes-align-to-work.html). Hope that makes it easier to review this.

**Comment by alexcrichton** [comment]
I, like @SimonSapin, have no idea what these functions are doing and find it pretty hard to follow...

This is a lang item though and doesn't really require libs team review, it seems like it's just an implementation detail. @nagisa do you know someone else who's familiar with this who can review?

**Comment by nagisa** [comment]
The only reason this is a language item is to make miri understand what kind of function this is (since miri has to handle raw pointers specially). Otherwise, it is just an implementation detail for the currently unstable library function `pointer::align_offset`, and, I believe, should fall into @rust-lang/libs purview.

Admittedly, this change is so minor that I’d be fine with *not* landing this if it ends up requiring somebody to spend a significant time in order to understand and review this code. I’ll leave the decision to close this PR up to you.

> do you know someone else who's familiar with this who can review?

Alas, the bus factor for this implementation is 2 (me and the person on IRC who collaborated with me in implementing this and whose nick I’ve forgotten by now). I had hoped to rectify this to some extent with my blog post (which I’ve linked in one of the comments above). If you are interested in raising the bus factor (I’m sure all of you have better things to do), I’d be willing to answer any questions as they come up.

Yet, I’d like to point out that understanding the whole algorithm is not necessary to review this PR. I already wrote about what the essence of this specific pull request does [here](https://github.com/rust-lang/rust/pull/54534#issuecomment-424335539).

I’d like to also point out that this algorithm is tested by a [fairly exhaustive test suite](https://github.com/rust-lang/rust/blob/e1041c6cd1a027ab3ada3e8538620d2e1d7067fe/src/libcore/tests/ptr.rs#L300-L387). These tests were the basis for accepting the initial code. I’m very confident that these tests would have caught any issues with the changes I’ve made in this PR.

**Comment by alexcrichton** [comment]
Ok well it sounds like you're pretty confident in these changes as well as the test coverage, and if it's primarily for an unstable API then it shouldn't be harmful to add in as well.

@bors: r+

**Comment by RalfJung** [comment]
That's a nice blog post! Might be worth adding a link to it from the code? (Keep in mind though that the post might disappear before the code does.)

**Comment by RalfJung** [comment]
Looks very spurious to me.

@bors retry
